src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -223,33 +225,59 @@ void CoreWorkerDirectActorTaskSubmitter::SendPendingTasks(const ActorID &actor_i
   if (!it->second.rpc_client) {
     return;
   }
+  auto &client_queue = it->second;
 
   // Check if there is a pending force kill. If there is, send it and disconnect the
   // client.
-  if (it->second.pending_force_kill) {
+  if (client_queue.pending_force_kill) {
     RAY_LOG(INFO) << "Sending KillActor request to actor " << actor_id;
     // It's okay if this fails because this means the worker is already dead.
-    it->second.rpc_client->KillActor(*it->second.pending_force_kill, nullptr);
-    it->second.pending_force_kill.reset();
+    client_queue.rpc_client->KillActor(*client_queue.pending_force_kill, nullptr);
+    client_queue.pending_force_kill.reset();
   }
 
   // Submit all pending requests.
-  auto &requests = it->second.requests;
+  auto &requests = client_queue.requests;
   auto head = requests.begin();
-  while (head != requests.end() && head->first <= it->second.next_send_position &&
-         head->second.second) {
+  while (head != requests.end() &&
+         (/*seqno*/ head->first <= client_queue.next_send_position) &&
+         (/*dependencies_resolved*/ head->second.second)) {
     // If the task has been sent before, skip the other tasks in the send
     // queue.
-    bool skip_queue = head->first < it->second.next_send_position;
+    bool skip_queue = head->first < client_queue.next_send_position;
     auto task_spec = std::move(head->second.first);
     head = requests.erase(head);
 
-    RAY_CHECK(!it->second.worker_id.empty());
-    PushActorTask(it->second, task_spec, skip_queue);
-    it->second.next_send_position++;
+    RAY_CHECK(!client_queue.worker_id.empty());
+    PushActorTask(client_queue, task_spec, skip_queue);
+    client_queue.next_send_position++;
   }
 }
 
+void CoreWorkerDirectActorTaskSubmitter::ResendOutOfOrderTasks(const ActorID &actor_id) {
+  auto it = client_queues_.find(actor_id);
+  RAY_CHECK(it != client_queues_.end());
+  if (!it->second.rpc_client) {
+    return;
+  }
+  auto &client_queue = it->second;
+
+  for (const auto &completed_task : client_queue.out_of_order_completed_tasks) {
+    // Making an copy here because we are flipping a flag and the original value is
+    // const.
+    TaskSpecification task_spec = completed_task.second;
+    // Calling SubmitTask in event loop to avoid deadlock condition. SubmitTask holds
+    // the mu_ and calls this method, we can't call SubmitTask recursively.
+    io_service_.post([this, task_spec]() mutable {
```

Comment:
Unless I'm mistaken, I believe this is making two copies -> one when you do the assignment operator and then a second when you capture it in the `io_service_.post` call. We should only make the second copy.

Reply:
Tried but turns out we can't do it in C++11 https://stackoverflow.com/questions/24221061/lambda-capture-list-capturing-objects-member-field-by-value-not-possible-witho

This is because if we just pass the first task_spec without copying, compiler complains because it's const (yes even when it's mutable), so we have to copy it to de-const it.